### PR TITLE
Add mod manager 

### LINKED ISSUE #1566: [Feature Request] Mod Manager
The only thing that doesn't have a manager now is mods so whenever you wanna disable something you need to either move it or delete it and put it back in if you wanna enable it.
Having an option to make it togleable like DLC and Updates would be nice.

### PR DESCRIPTION
**Pull request by Caian**
This PR replaces the "Open mod directory" option with a mod manager window that allows inspecting mods and selecting which ones will be activated when the game starts. It also allows quick access to the per-game and global mod directories, as well as the directory of each individual mod.

ModManager

It can detect that a mod previously enabled has been removed and warn the user about it:

mods2

There are also additional log messages to indicate excluded mods when starting a game:

```
00:00:17.189 |I| ModLoader FilterMods: Mod <60 FPS Cheat> is not enabled
00:00:17.189 |I| ModLoader FilterMods: Mod <30 FPS Cheat> is not enabled
00:00:17.189 |I| ModLoader FilterMods: Mod <60 FPS Cheat> is not enabled
```

Closes #1566.

### CODE REVIEW

**Comment by Mou-Ikkai** [comment]
Seems to work! Tried it with a Megamix mod that added new modules (outfits) to the game.
image
image
image
image
(You can see the Alien module is now gone and it defaulted back to regular Miku after mod was disabled.)

**Comment by Myoko** [comment]
lgtm!

File: Ryujinx.HLE/HOS/ModLoader.cs [hunk 1]
```diff
@@ -421,6 +421,41 @@ public void CollectMods(IEnumerable<ulong> titles, params string[] searchDirPath
...
+            foreach (ulong titleId in titles)
+            {
+                ModCache unfilteredModCache = allAppMods[titleId];
+                ModCache filteredModCache = new ModCache();
+
+                FilterMods(unfilteredModCache.RomfsContainers, filteredModCache.RomfsContainers, enabledMods);
+                FilterMods(unfilteredModCache.ExefsContainers, filteredModCache.ExefsContainers, enabledMods);
+                FilterMods(unfilteredModCache.RomfsDirs      , filteredModCache.RomfsDirs      , enabledMods);
+                FilterMods(unfilteredModCache.ExefsDirs      , filteredModCache.ExefsDirs      , enabledMods);
+                FilterMods(unfilteredModCache.Cheats         , filteredModCache.Cheats         , enabledMods);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/ModLoader.cs:453]:
```suggestion
                FilterMods(unfilteredModCache.RomfsContainers, filteredModCache.RomfsContainers, enabledMods);
                FilterMods(unfilteredModCache.ExefsContainers, filteredModCache.ExefsContainers, enabledMods);
                FilterMods(unfilteredModCache.RomfsDirs,       filteredModCache.RomfsDirs,       enabledMods);
                FilterMods(unfilteredModCache.ExefsDirs,       filteredModCache.ExefsDirs,       enabledMods);
                FilterMods(unfilteredModCache.Cheats,          filteredModCache.Cheats,          enabledMods);
```

File: Ryujinx/Ui/Widgets/GameTableContextMenu.cs [hunk 2]
```diff
@@ -540,9 +537,9 @@ private void PurgePtcCache_Clicked(object sender, EventArgs args)
...
                 foreach (FileInfo file in cacheFiles)
                 {
-                    try 
-                    { 
-                        file.Delete(); 
+                    try
+                    {
+                        file.Delete();
                     }
                     catch(Exception e)
```
Comment (AcK77) [inline Ryujinx/Ui/Widgets/GameTableContextMenu.cs:544]:
```suggestion
                    catch (Exception e)
```

File: Ryujinx/Ui/Windows/ModsWindow.cs [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            _modsTreeView.AppendColumn("Enabled", enableToggle,           "active", 0);
+            _modsTreeView.AppendColumn("Name",    new CellRendererText(), "text",   1);
+            _modsTreeView.AppendColumn("Type",    new CellRendererText(), "text",   2);
+            _modsTreeView.AppendColumn("Path",    new CellRendererText(), "text",   3);
+            _modsTreeView.ButtonReleaseEvent += Row_Clicked;
+
+            Refresh();
+        }
+
+        public static string GetModsJsonPath(string titleId)
```
Comment (AcK77) [inline Ryujinx/Ui/Windows/ModsWindow.cs:74]:
```suggestion
        private static string GetModsJsonPath(string titleId)
```

**Review by AcK77 — CHANGES_REQUESTED** [review]

**Comment by MrRobot-o** [comment]
Great job with this mod manager PR.
The mod manager can also find mods which are located in DLC folders.
I tried "Relics of the past" mod and it seems to be working fine as both parts show up.

**Comment by GamerzHell9137** [comment]
Closes https://github.com/Ryujinx/Ryujinx/issues/1566

**Comment by GhostofReachCashGrab** [comment]
Found no issues with this Pr with any Mods or cheats used seems like it's ready to go nice work!

**Comment by Caian** [comment]
Add warning when a previously enabled mod is removed .

**Comment by Torphedo** [comment]
Is there anything left to do here? This seems like a cool PR that got left in the dust a bit. Would love to see this merged!

**Comment by Caian** [comment]
This PR is done, aside from fixing eventual bugs reported by the community and following the recomendations of other maintainers.

It is to be decided though if the PR should be merged before or after #2507, as it introduces more graphical modifications that must be ported.

**Comment by Zero17N5** [comment]
Hi, sorry I am new here, How do I download this and how do I instal it?

**Comment by HeatPhoenix** [comment]
Is there any reason why this was never merged? Looked very useful to me.

**Comment by MutantAura** [comment]
The author does not intend to work on this further.

**Pull request closed without merging**